titan-server/src/api.rs

```diff
@@ -58,73 +60,104 @@ impl<'r, T: Serialize> Responder<'r> for ApiResponse<T> {
             Ok(res) => {
                 json!(res).respond_to(req)
             },
-            Err(err) => err.respond_to(req)
+            Err(err) => {
+                err.respond_to(req)
+            }
         }
     }
 }
 
 /// Fields returned from the API when an error occurs.
 pub struct ApiErrorResponse {
-    body: ApiResponseBody,
+    body: ApiErrorResponseBody,
     status: Status,
 }
 
-pub type ApiErrorResponseResult = Result<ApiErrorResponse, Status>;
+pub type ApiErrorResponseResult<'r> = Result<ApiErrorResponse, Status>;
 
-impl From<TitanDatabaseError> for ApiErrorResponseResult {
+impl<'r> From<TitanDatabaseError> for ApiErrorResponseResult<'r> {
     fn from(err: TitanDatabaseError) -> Self {
         match err {
             TitanDatabaseError::DieselError(db_error) => {
                 if db_error == DieselError::NotFound {
                     return Err(Status::NotFound);
                 }
-                Err(Status::InternalServerError)
+
+                Ok(ApiErrorResponse {
+                    body: ApiErrorResponseBody {
+                        code: 500,
+                        description: "Internal server error".to_owned(),
+                        message: Some(db_error.description().to_owned()),
+                    },
+                    status: Status::InternalServerError
+                })
             },
             _ => Err(Status::InternalServerError),
         }
     }
 }
 
-#[derive(Serialize)]
-pub struct ApiResponseBody {
+#[derive(Debug, Serialize)]
+pub struct ApiErrorResponseBody {
+    /// HTTP status code
+    pub code: i16,
+    /// HTTP status description
+    pub description: String,
     /// Describes the cause of an error. The contents of this message
     /// should be easily understood by the user and not expose
     /// sensitive details that may violate the security of the user
     /// or system.
-    pub message: String,
+    pub message: Option<String>,
 }
 
 /// Enumeration of error types that may be returned in a response
 /// to client apps.
 pub enum ApiError {
     AuthenticationError,
     TitanDatabaseError(TitanDatabaseError),
+    ValidationError(&'static str),
 }
 
 impl<'r> Responder<'r> for ApiError {
     /// Implementation of Rocket's `Responder`. Converts API errors to
     /// a user-friendly JSON structure.
-    fn respond_to(self, _: &Request) -> Result<Response<'r>, Status> {
+    fn respond_to(self, req: &Request) -> Result<Response<'r>, Status> {
```

Comment:
The `req` variable appears to be unused. We should prefix it (or replace it) with an underscore to avoid compiler warnings.